_posts/2017-10-06-rust-blog-series-2.md

```diff
@@ -0,0 +1,1098 @@
+The data structs for our models *may* look like this.
+- User
+- NewUser
+- Post
+- NewPost
+
+Lets create a `src/models.rs` file and review the code that goes into it.
+
+
+```rust
+// Inside `src/models.rs`
+
+// We JUST made the schema file...
+// Lets take advantage of it by bringing it into scope here
+use schema::{posts, users};
+
+#[derive(Debug, Queryable)]
+struct User {
+    id: i32,
+    first_name: String,
+    last_name: String,
+    email: String,
+    password: String, 
+}
+
+#[derive(Debug, Insertable)]
+#[table_name="users"]
+struct NewUser {
+    first_name: String,
+    last_name: String,
+    email: String,
+    password: String, 
+}
+
+#[derive(Debug, Queryable)]
+struct Post {
+    id: i32,
+    user_id: i32,
+    title: String,
+    content: String,
+    published: bool,
+}
+
+#[derive(Debug, Insertable)]
+#[table_name="posts"]
+struct NewPost {
+    user_id: i32,
+    title: String,
+    content: String,
+}
+
+```
+
+If you're thinking *"whoa, whoa, whoa... WUT are those weird DERIVE things?!?!"*,
+don't worry, they're your friend!
+
+The `diesel_codegen` crate... well, generates code for us!
+Rust is strongly typed and has a wonderful trait system which Diesel takes advantage of.
+This gives us compile-time guarantees about all aspects of our models.
+
+`Debug` should look somewhat familiar, as it's basically standard with the Rust langauge.
+The really interesting ones are `Queryable` and `Insertable`.
+`Insertable` is Diesel's way of saying that *the values of this struct map to the columns of a table and can be inserted in a row*
+With `Insertable`, we don't generally include any auto-incrementing primary key columns because our database takes care of that for us.
+
+`Queryable` is also self explanatory.
+A `Queryable` struct is one that may be queried from the database with those fields / columns.
+Our `Queryable` structs share the same name as the table, so the table name is inferred,
+however our `Insertable` structs do not, so we must annotate them with the table they correspond to.
+Our current code is very basic and we'll just select all columns for now when querying.
+
+Diesel supports nearly all database column types and maps them to Rust values.
+We're going to primarily be using `i32`, and `String`, but you'll see we will eventually use
+`&str`, and datetime types from the `chrono` crate.
+Diesel supports custom types, but this requires some impls on your part.
+We will not be covering advanced cases like that in this blog series.
+
+One small thing to note is our `NewPost` struct is missing the `published` field.
+This is because it has a default value of `FALSE`.
+Thinking ahead a little, we will want a feature of the app to publish posts after they are done.
+This is one way of keeping posts in a *draft* state.
+[This document] covers all the model derives in much more detail.
+
+[This document]: https://github.com/diesel-rs/diesel/blob/master/guide_drafts/model-derives.md
```

Comment:
Do you think it's worth holding off on linking to guide drafts until after they've been published?